### ISSUE TITLE
[TODO] [Day 8] Two-Stage Classifier 구현

### ISSUE THREAD

**Issue by nimnusmik** [AIBootcamp14/computervisioncompetition-cv-1#40]
# 📋 작업 개요
**작업 주제:** EfficientNet 모델 최적화 - val_f1 0.95 이상 달성  
**담당자:** @nimnusmik  
**마감일:** 2025-09-08 24:00  
**예상 소요시간:** 총 1-2일 (GPU 환경 기준)

## 📅 기간
- 시작일: 2025-09-08  
- 종료일: 2025-09-10

## 📌 목표
- EfficientNet-B3 기반 모델의 val_f1 스코어를 0.95 이상으로 개선  
- 문제 클래스(3, 7, 14)의 F1 스코어를 0.7 이상으로 끌어올리기  
- 데이터 불균형과 과적합 문제를 해결하여 모델 성능 안정화  
- WandB를 활용한 실험 추적 및 최적 하이퍼파라미터 도출

## ✅ Todo List
- [x] 클래스 분포 분석 및 문제 클래스(3, 7, 14) 데이터 확인  
- [ ] Focal Loss 및 클래스 가중치 적용으로 불균형 해소  
- [x] Augmentation 전략 강화 (문제 클래스 전용 transform 추가)  
- [ ] Learning Rate Scheduler를 OneCycleLR로 변경  
- [ ] Optuna를 활용한 하이퍼파라미터 튜닝 (LR, gamma, mixup_alpha)  
- [ ] EfficientNet-B5 또는 ConvNeXt로 모델 업그레이드 및 앙상블  
- [ ] TTA 확장 및 최종 제출 파일 생성  

### 핵심 아이디어: 단계별 성능 개선
- **1단계**: 기본 튜닝 (Focal Loss, OneCycleLR, Patience 증가)으로 val_f1 0.93~0.94  
- **2단계**: 문제 클래스 타겟팅 (오버샘플링, 강한 augmentation)으로 F1 0.7+  
- **3단계**: 모델 업그레이드 (EfficientNet-B5, 앙상블) 및 Optuna로 val_f1 0.95 돌파  
- 예시: 클래스 3, 7, 14의 F1이 낮은 이유는 데이터 부족 또는 이미지 왜곡 가능성. 전용 augmentation으로 해결.

### 유용한 링크
- [ ] EfficientNet 논문: [[EfficientNet: Rethinking Model Scaling](https://arxiv.org/abs/1905.11946)](https://arxiv.org/abs/1905.11946)  
- [ ] Timm 라이브러리 문서: [https://timm.fast.ai/](https://timm.fast.ai/)  
- [ ] Focal Loss 논문: [[Focal Loss for Dense Object Detection](https://arxiv.org/abs/1708.02002)](https://arxiv.org/abs/1708.02002)  
- [ ] Optuna 문서: [https://optuna.org/](https://optuna.org/)  

---

### 실험 중 발견사항
- **클래스 불균형**: 문제 클래스(3, 7, 14)의 데이터 비율이 10% 미만일 가능성. 오버샘플링 필요.  
- **과적합**: Train F1 0.7~0.8 vs Val F1 0.9. Label Smoothing 및 Dropout 강화.  
- **문제 클래스 이미지**: 희미한 텍스트/왜곡 가능성. RandomBrightnessContrast, ElasticTransform 추가.  

### 다음 실험 아이디어
- **Hierarchical Augmentation**: 문제 클래스에만 강한 노이즈/왜곡 적용  
- **Ensemble + TTA**: EfficientNet-B5와 ConvNeXt 앙상블 후 TTA 10개로 확률 보정  
- **Feature-level Regularization**: EfficientNet head에 Dropout(0.3) 추가  
- **Confidence Calibration**: Platt Scaling으로 예측 확률 보정  
- **Data Inspection**: 클래스 14의 라벨링 오류 점검  

### 팀원 피드백
- **성능 향상 목표**: val_f1 0.95 이상, 문제 클래스 F1 0.7 이상  
- **추론 속도 제한**: TTA 증가 시 추론 시간 1.5배 이하 유지  
- **메모리 사용량**: EfficientNet-B5 사용 시 GPU 메모리 모니터링 (torch.cuda.empty_cache 활용)  

---

 
### 성능 측정 지표
- **전체 성능**: val_f1 (목표: 0.95 이상)  
- **클래스별 성능**: 문제 클래스(3, 7, 14)의 F1-score (목표: 0.7 이상)  
- **추론 속도**: 평균 샘플당 추론 시간 (목표: 기존 대비 1.5배 이하)  
- **메모리 사용량**: 피크 GPU 메모리 사용량 (WandB로 모니터링)

**Comment by nimnusmik** [comment]
# step 1. 일단 간단하게 손실함수 변경하기 

- 기존: `loss_fn = nn.CrossEntropyLoss(label_smoothing=0.2)`
- 현재: ```class_weights = torch.FloatTensor([
    2.5 if i in [3, 7, 14] else 1.0 for i in range(17)
]).to(device)
loss_fn = nn.CrossEntropyLoss(weight=class_weights, label_smoothing=0.15)```
	- 클래스 3, 7, 14는 가중치를 2.5로 줌 → 모델이 이 클래스에 더 집중하게 됨.
	- 나머지 클래스는 1.0 → 기본 가중치.

- 각 fold에서 클래스별 성능을 추적하여 개선 효과를 확인

<img width="1030" height="735" alt="Image" src="https://github.com/user-attachments/assets/7f23bfbf-1ee3-4719-a28e-d52772c9935d" />

## 3,7,14에 대한 수치가 확실히 현저히 낮다. 이부분을 개선해야할 필요가 있음

<img width="963" height="421" alt="Image" src="https://github.com/user-attachments/assets/55297125-b991-4a99-9c37-58760f407b8c" />

3,7,14 클래스를 잘 맞추면 전체 F1이 올라간다. 

## 추가 step
-  가중치 2.5 -> 3 : 유의미한 변화 X
-  가중치 = 4& LR = 2e-4 조정: 성능 더 떨어짐 (0.92 -> 0.90)


## 현재 문제점 분석

전체 F1은 높지만(0.9285) 문제 클래스(3,7,14) 성능이 지속적으로 낮음(0.55-0.61)
27에폭까지 학습했는데도 문제 클래스 성능이 개선되지 않음
패턴을 보면 구조적 문제로 보임 (단순히 더 오래 학습한다고 해결될 문제가 아님)


# Step 2. 문제 클래스 전용 augmentation 도임

**Comment by nimnusmik** [comment]
또 태워먹음 서버ㅠㅜㅜㅠㅜㅠ푸ㅠ푸ㅠㅜㅠㅜㅠㅠㅜㅜㅠㅜㅠㅜ

하ㅏㅏ가간ㄱㅁ
